### PR TITLE
Add no-unused-styles rule

### LINKED ISSUE #2: Rule proposal: no unused styles
All of the styles for a component should be self-contained within the component. This should make it pretty easy to identify styles that are defined but not actually used by the component, which would be a useful rule to have.

Bad:

``` jsx
function MyComponent({ styles }) {
  return (
    <div {...css(styles.foo)}>
      Foo
    </div>
  );
}

export default withStyles(() => ({
  foo: {
    backgroundColor: 'red',
  },

  bar: { // <--- this style is not used
    backgroundColor: 'green',
  },
}))(MyComponent);
```

Good:

``` jsx
function MyComponent({ styles }) {
  return (
    <div {...css(styles.foo)}>
      Foo
    </div>
  );
}

export default withStyles(() => ({
  foo: {
    backgroundColor: 'red',
  },
}))(MyComponent);
```

### PR DESCRIPTION
**Pull request by lencioni**
Fixes #2.

@airbnb/webinfra

### CODE REVIEW

File: lib/util/findImportCSSFromWithStylesImportDeclaration.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  if (!node.source.value.endsWith('withStyles')) {
+    return null;
+  }
+
+  if (!node.specifiers.length) {
+    // import 'foo';
+    return null;
+  }
+
+  let name = null;
```
Comment (ljharb) [inline lib/util/findImportCSSFromWithStylesImportDeclaration.js]: since this isn't being transpiled, the file needs to be explicitly in strict mode

Comment (lencioni) [inline lib/util/findImportCSSFromWithStylesImportDeclaration.js]: Done!

**Comment by lencioni** [comment]
> I'm concerned with the number of caveats/limitations in the readme

Not ideal, but perfect is the enemy of the good.

> Is there a reason not to cover "Will not detect styles used by Literal reference"? String literals, and template literals with no interpolations, should be just as statically detectable as dot notation.

Mostly I ran out of time yesterday. Also, I think the value is pretty low in our codebase at least since it is unlikely someone would do `styles['foo']` instead of `styles.foo`. I'm comfortable shipping and improving later.

**Comment by ljharb** [comment]
@lencioni i'm concerned about things that aren't valid identifiers. Namely, `styles[':first-child']` and so forth.

**Comment by lencioni** [comment]
That will just end up being `undefined`, unless the key `:first-child` exists, in which case it should work just fine. I think the rest of the problems would be covered by no-undefined-styles in #3, or some other rule that warned about silliness like `styles[':first-child']`.

**Comment by ljharb** [comment]
Why are pseudo-selectors silliness?

**Comment by ljharb** [comment]
Or do you mean that the styles property names wouldn't be doing that, so it doesn't apply?

**Comment by lencioni** [comment]
`styles[':first-child']` looks like a pseudo selector, but I don't believe it will be treated that way. But I suppose that is sorta up to the underlying engine. It looks like Aphrodite will generate a className like the following with such a strange style property: `:first-child_im3wl1`. Of course this is invalid, so perhaps this is a bug on the Aphrodite side of things, or maybe just an edge case they don't think is worth bothering about.

If you wanted a pseudo-selector like this, you would need to define it like:

```js
foo: {
  ':first-child': {
    ...
  }
}
```

And apply it via `css(styles.foo)`.

Also, for CSS in JS, pseudo-selectors should generally be avoided in favor of explicit logic that does not implicitly affect global state.

**Pull request merged**